File: primer_corte/Prototype/src/Email.java
// Prototipo concreto: Email
class Email extends AbstractDocument {
    private String recipients;
    private String subject;
    private String body;
    private boolean hasAttachments;
    
    public Email() {
    }
    
    // Constructor prototipo
    public Email(Email source) {
        // Llamar al constructor padre para copiar los campos de la clase padre
        super(source);
        if (source != null) {
            this.recipients = source.recipients;
            this.subject = source.subject;
            this.body = source.body;
            this.hasAttachments = source.hasAttachments;
        }
    }
    
    // Getters y setters
    public String getRecipients() {
        return recipients;
    }
    
    public void setRecipients(String recipients) {
        this.recipients = recipients;
    }
    
    public String getSubject() {
        return subject;
    }
    
    public void setSubject(String subject) {
        this.subject = subject;
    }
    
    public String getBody() {
        return body;
    }
    
    public void setBody(String body) {
        this.body = body;
    }
    
    public boolean isHasAttachments() {
        return hasAttachments;
    }
    
    public void setHasAttachments(boolean hasAttachments) {
        this.hasAttachments = hasAttachments;
    }
    
    @Override
    public Document clone() {
        return new Email(this);
    }
    
    @Override
    public void print() {
        System.out.println("---- EMAIL ----");
        System.out.println("Title: " + getTitle());
        System.out.println("Author: " + getAuthor());
        System.out.println("Creation Date: " + getCreationDate());
        System.out.println("To: " + recipients);
        System.out.println("Subject: " + subject);
        System.out.println("Body: " + body);
        System.out.println("Has Attachments: " + (hasAttachments ? "Yes" : "No"));
    }
    
    @Override
    public String toString() {
        return "Email [title=" + getTitle() + ", subject=" + subject + 
               ", recipients=" + recipients + ", hasAttachments=" + hasAttachments + "]";
    }
}

File: primer_corte/PrototypePython/email.py
from abstract_document import AbstractDocument

# Prototipo concreto: Email
class Email(AbstractDocument):
    def __init__(self, source=None):
        self.recipients = None
        self.subject = None
        self.body = None
        self.has_attachments = False
        
        # Inicializar atributos de la clase padre
        super().__init__()
        
        # Si se proporciona una fuente, copiar sus atributos
        if source:
            self.__init_copy__(source)
            self.recipients = source.recipients
            self.subject = source.subject
            self.body = source.body
            self.has_attachments = source.has_attachments

    # Getters y setters
    def get_recipients(self):
        return self.recipients

    def set_recipients(self, recipients):
        self.recipients = recipients

    def get_subject(self):
        return self.subject

    def set_subject(self, subject):
        self.subject = subject

    def get_body(self):
        return self.body

    def set_body(self, body):
        self.body = body

    def is_has_attachments(self):
        return self.has_attachments

    def set_has_attachments(self, has_attachments):
        self.has_attachments = has_attachments

    def clone(self):
        return Email(self)

    def print(self):
        print("---- EMAIL ----")
        print(f"Titulo: {self.get_title()}")
        print(f"Autor: {self.get_author()}")
        print(f"Fecha de creacion: {self.get_creation_date()}")
        print(f"Para: {self.recipients}")
        print(f"Asunto: {self.subject}")
        print(f"Cuerpo: {self.body}")
        print(f"Anexos: {'Yes' if self.has_attachments else 'No'}")
        print("--------------")

    def __str__(self):
        return f"Email [title={self.get_title()}, subject={self.subject}, " \
               f"recipients={self.recipients}, hasAttachments={self.has_attachments}]"
